demo_nodes_cpp/src/events/matched_event_detect.cpp

```diff
@@ -0,0 +1,223 @@
+// of MatchedEventDetectNode.
+
+class MatchedEventDetectNode : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC explicit MatchedEventDetectNode(
+    const std::string & pub_topic_name,
+    const std::string & sub_topic_name)
+  : Node("matched_event_detect_node")
+  {
+    rclcpp::PublisherOptions pub_options;
+    pub_options.event_callbacks.matched_callback =
+      [this](rclcpp::MatchedInfo & s) {
+        if (any_subscription_connected_) {
+          if (s.current_count == 0) {
+            RCLCPP_INFO(this->get_logger(), "Last subscription is disconnected.");
+            any_subscription_connected_ = false;
+          } else {
+            RCLCPP_INFO(
+              this->get_logger(),
+              "The changed number of connected subscription is %d and current number of connected"
+              " subscription is %lu.", s.current_count_change, s.current_count);
+          }
+        } else {
+          if (s.current_count != 0) {
+            RCLCPP_INFO(this->get_logger(), "First subscription is connected.");
+            any_subscription_connected_ = true;
+          }
+        }
+      };
+
+    pub_ = create_publisher<std_msgs::msg::String>(
+      pub_topic_name, 10, pub_options);
+
+    rclcpp::SubscriptionOptions sub_options;
+    sub_options.event_callbacks.matched_callback =
+      [this](rclcpp::MatchedInfo & s) {
+        if (any_publisher_connected_) {
+          if (s.current_count == 0) {
+            RCLCPP_INFO(this->get_logger(), "Last publisher is disconnected.");
+            any_publisher_connected_ = false;
+          } else {
+            RCLCPP_INFO(
+              this->get_logger(),
+              "The changed number of connected publisher is %d and  Current number of connected"
+              " publisher is %lu.", s.current_count_change, s.current_count);
+          }
+        } else {
+          if (s.current_count != 0) {
+            RCLCPP_INFO(this->get_logger(), "First publisher is connected.");
+            any_publisher_connected_ = true;
+          }
+        }
+      };
+    sub_ = create_subscription<std_msgs::msg::String>(
+      sub_topic_name,
+      10,
+      [](std_msgs::msg::String::ConstSharedPtr) {},
+      sub_options);
+  }
+
+private:
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_;
+  rclcpp::Subscription<std_msgs::msg::String>::SharedPtr sub_;
+  bool any_subscription_connected_{false};
+  bool any_publisher_connected_{false};
+};
+
+class MultiSubNode : public rclcpp::Node
+{
+public:
+  DEMO_NODES_CPP_PUBLIC explicit MultiSubNode(const std::string & topic_name)
+  : Node("multi_sub_node"),
+    topic_name_(topic_name)
+  {}
+
+  DEMO_NODES_CPP_PUBLIC void create_one_sub(void)
+  {
+    auto sub = create_subscription<std_msgs::msg::String>(
+      topic_name_,
+      10,
+      [](std_msgs::msg::String::ConstSharedPtr) {});
+    subs_.emplace_back(sub);
+  }
+
+  DEMO_NODES_CPP_PUBLIC void destroy_one_sub(void)
```

Comment:
Even if it's a demo that we can control the whole process, I think we should design a better behavior in this method. It's just a suggestion,
```
rclcpp::Subscription<std_msgs::msg::String>::WeakPtr create_one_sub(void);
void destroy_one_sub(rclcpp::Subscription<std_msgs::msg::String>::WeakPtr sub);
```

Reply:
Do you means test as the below
```c++
  auto sub1 = multi_sub_node->create_one_sub();
  executor.spin_some(200ms);

  auto sub2 = multi_sub_node->create_one_sub();
  executor.spin_some(200ms);

  multi_sub_node->destroy_one_sub(sub1);
  executor.spin_some(200ms);

  multi_sub_node->destroy_one_sub(sub2);
  executor.spin_some(200ms);
```

Reply:
Yes.